src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp

```diff
@@ -28,249 +29,847 @@
+  assert (idx < _max, "index is sane: " SIZE_FORMAT " < " SIZE_FORMAT, idx, _max);
+  assert (membership(idx) == ShenandoahFreeSetPartitionId::NotFree, "Cannot make free if already free");
+  assert (int(which_partition) < NumPartitions, "selected free partition must be valid");
+  assert (available <= _region_size_bytes, "Available cannot exceed region size");
+
+  _membership[int(which_partition)].set_bit(idx);
+  _capacity[int(which_partition)] += _region_size_bytes;
+  _used[int(which_partition)] += _region_size_bytes - available;
+  expand_interval_if_boundary_modified(which_partition, idx, available);
+
+  _region_counts[int(which_partition)]++;
+}
+
+void ShenandoahRegionPartitions::move_from_partition_to_partition(idx_t idx, ShenandoahFreeSetPartitionId orig_partition,
+                                                                  ShenandoahFreeSetPartitionId new_partition, size_t available) {
+  assert (idx < _max, "index is sane: " SIZE_FORMAT " < " SIZE_FORMAT, idx, _max);
+  assert (int(orig_partition) < NumPartitions, "Original partition must be valid");
+  assert (int(new_partition) < NumPartitions, "New partition must be valid");
+  assert (available <= _region_size_bytes, "Available cannot exceed region size");
+
+  // Expected transitions:
+  //  During rebuild:         Mutator => Collector
+  //  During flip_to_gc:      Mutator empty => Collector
+  // At start of update refs: Collector => Mutator
+  assert (((available <= _region_size_bytes) &&
+           (((orig_partition == ShenandoahFreeSetPartitionId::Mutator)
+             && (new_partition == ShenandoahFreeSetPartitionId::Collector)) ||
+            ((orig_partition == ShenandoahFreeSetPartitionId::Collector)
+             && (new_partition == ShenandoahFreeSetPartitionId::Mutator)))) ||
+          ((available == _region_size_bytes) &&
+           ((orig_partition == ShenandoahFreeSetPartitionId::Mutator)
+            && (new_partition == ShenandoahFreeSetPartitionId::Collector))), "Unexpected movement between partitions");
+
+  size_t used = _region_size_bytes - available;
+
+  _membership[int(orig_partition)].clear_bit(idx);
+  _membership[int(new_partition)].set_bit(idx);
+
+  _capacity[int(orig_partition)] -= _region_size_bytes;
+  _used[int(orig_partition)] -= used;
+  shrink_interval_if_boundary_modified(orig_partition, idx);
+
+  _capacity[int(new_partition)] += _region_size_bytes;;
+  _used[int(new_partition)] += used;
+  expand_interval_if_boundary_modified(new_partition, idx, available);
+
+  _region_counts[int(orig_partition)]--;
+  _region_counts[int(new_partition)]++;
+}
+
+const char* ShenandoahRegionPartitions::partition_membership_name(idx_t idx) const {
+  assert (idx < _max, "index is sane: " SIZE_FORMAT " < " SIZE_FORMAT, idx, _max);
+  ShenandoahFreeSetPartitionId result = ShenandoahFreeSetPartitionId::NotFree;
+  for (uint partition_id = 0; partition_id < NumPartitions; partition_id++) {
```

Comment:
Looks to me like you could just call the below membership() method (and make membership() available outside of ASSERT), and then pass the result to partition_name()?

Reply:
Thanks.  I'll make this change.